datafusion-physical-expr/src/conditional_expressions.rs

```diff
@@ -0,0 +1,84 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::sync::Arc;
+
+use arrow::array::Array;
+use arrow::datatypes::DataType;
+
+use datafusion_common::{DataFusionError, Result, ScalarValue};
+use datafusion_expr::ColumnarValue;
+
+pub fn coalesce(args: &[ColumnarValue]) -> Result<ColumnarValue> {
+    // do not accept 0 arguments.
+    if args.is_empty() {
+        return Err(DataFusionError::Internal(format!(
+            "coalesce was called with {} arguments. It requires at least 1.",
+            args.len()
+        )));
+    }
+
+    let mut res = vec![];
+    let size = match args[0] {
+        ColumnarValue::Array(ref a) => a.len(),
+        ColumnarValue::Scalar(ref _s) => 1,
+    };
+
+    for i in 0..size {
+        let mut value = ScalarValue::try_from(&args[0].data_type())?;
+        for column_value in args {
+            match column_value {
+                ColumnarValue::Array(array_ref) => {
+                    if array_ref.is_valid(i) {
```

Comment:
This implementation can likely be optimized (I think it should be >10x faster) by not converting values to `ScalarValue` and using `is_valid`, matching per value, but operating directly on Arrays. That would also allow optimizations like skipping running the coalesce althogether if all items in the array already are non null, or skipping handling the array if every value is null, etc.

This will involve some more work / code though, but maybe a comment that there are some big optimizations possible could be useful.

The current implementation also could likely already be optimized a bit by moving the for loop to the innermost place instead of the outermost place and/or implementing one of the rules like mentioned.